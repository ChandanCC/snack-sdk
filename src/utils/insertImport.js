/* @flow */

import { types } from 'recast';

const n = types.namedTypes;
const b = types.builders;

type ImportDescription = {
  from: string,
  name?: string,
  default?: boolean,
};

const insertImport = (ast: any, item: ImportDescription) => {
  let imported = false;

  types.visit(ast, {
    visitImportDeclaration(path) {
      // Check if there's an import for our module
      if (path.node.source.value === item.from) {
        // Check if there's an import for specific export
        if (item.default) {
          imported = path.node.specifiers.some(n.ImportDefaultSpecifier.check);
        } else if (item.name) {
          imported = path.node.specifiers
            .filter(n.ImportSpecifier.check)
            .some(s => s.imported.name === item.name);
        } else {
          imported = true;
        }

        // There's an import for the module, but not the specific export
        if (!imported && item.name) {
          if (item.default) {
            path.node.specifiers.unshift(
              b.importDefaultSpecifier(b.identifier(item.name))
            );
          } else {
            path.node.specifiers.push(
              b.importSpecifier(b.identifier(item.name))
            );
          }
          imported = true;
        }
      }

      this.traverse(path);
    },
  });

  // There are no import statements for the module
  if (!imported) {
    // Find the index of last import statement
    const index = ast.program.body.reduce((acc, curr, i) => {
      if (n.ImportDeclaration.check(curr)) {
        return i;
      }
      return acc;
    }, 0);

    // Insert our import after the last import statement
    ast.program.body.splice(
      index + 1,
      0,
      item.name
        ? b.importDeclaration(
            item.default
              ? [b.importDefaultSpecifier(b.identifier(item.name))]
              : [b.importSpecifier(b.identifier(item.name))],
            b.literal(item.from)
          )
        : b.importDeclaration([], b.literal(item.from))
    );
  }
};

export default insertImport;
