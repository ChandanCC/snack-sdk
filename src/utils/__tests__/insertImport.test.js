/* eslint-env jest */

import { parse, print } from 'recast';
import insertImport from '../insertImport';
import * as babylon from 'babylon';

import config from '../../configs/babylon';

const parser = {
  parse: (code: string) => babylon.parse(code, config),
};

it('inserts default import for new module', () => {
  const code = `
    import React from 'react';

    export default () => <div>Hello world</div>;
  `;

  const ast = parse(code, { parser });

  insertImport(ast, {
    name: 'lodash',
    from: 'lodash',
    default: true,
  });

  expect(print(ast).code).toMatchSnapshot();
});

it('inserts named import for new module', () => {
  const code = `
    import React from 'react';

    export default () => <div>Hello world</div>;
  `;

  const ast = parse(code, { parser });

  insertImport(ast, {
    name: 'Components',
    from: 'exponent',
  });

  expect(print(ast).code).toMatchSnapshot();
});

it('inserts empty import for new module', () => {
  const code = `
    import React from 'react';

    export default () => <div>Hello world</div>;
  `;

  const ast = parse(code, { parser });

  insertImport(ast, {
    from: 'babel/polyfill',
  });

  expect(print(ast).code).toMatchSnapshot();
});

it('inserts default import for existing module', () => {
  const code = `
    import { Component } from 'react';

    export default () => <div>Hello world</div>;
  `;

  const ast = parse(code, { parser });

  insertImport(ast, {
    name: 'React',
    from: 'react',
    default: true,
  });

  expect(print(ast).code).toMatchSnapshot();
});

it('inserts named import for existing module', () => {
  const code = `
    import React from 'react';

    export default () => <div>Hello world</div>;
  `;

  const ast = parse(code, { parser });

  insertImport(ast, {
    name: 'Component',
    from: 'react',
  });

  expect(print(ast).code).toMatchSnapshot();
});

it('does not insert duplicate default import for existing module', () => {
  const code = `
    import React, { Component } from 'react';

    export default () => <div>Hello world</div>;
  `;

  const ast = parse(code, { parser });

  insertImport(ast, {
    name: 'React',
    from: 'react',
    default: true,
  });

  expect(print(ast).code).toMatchSnapshot();
});

it('does not insert duplicate named import for existing module', () => {
  const code = `
    import { Component } from 'react';

    export default () => <div>Hello world</div>;
  `;

  const ast = parse(code, { parser });

  insertImport(ast, {
    name: 'Component',
    from: 'react',
  });

  expect(print(ast).code).toMatchSnapshot();
});

it('does not insert duplicate empty import for new module', () => {
  const code = `
    import { Component } from 'react';
    import 'babel/polyfill';

    export default () => <div>Hello world</div>;
  `;

  const ast = parse(code, { parser });

  insertImport(ast, {
    from: 'babel/polyfill',
  });

  expect(print(ast).code).toMatchSnapshot();
});
