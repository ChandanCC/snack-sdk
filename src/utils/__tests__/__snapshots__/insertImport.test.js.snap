exports[`does not insert duplicate default import for existing module 1`] = `
"
    import React, { Component } from 'react';

    export default () => <div>Hello world</div>;
  "
`;

exports[`does not insert duplicate empty import for new module 1`] = `
"
    import { Component } from 'react';
    import 'babel/polyfill';

    export default () => <div>Hello world</div>;
  "
`;

exports[`does not insert duplicate named import for existing module 1`] = `
"
    import { Component } from 'react';

    export default () => <div>Hello world</div>;
  "
`;

exports[`inserts default import for existing module 1`] = `
"
    import React, { Component } from 'react';

    export default () => <div>Hello world</div>;
  "
`;

exports[`inserts default import for new module 1`] = `
"import React from 'react';

import lodash from "lodash";

export default () => <div>Hello world</div>;"
`;

exports[`inserts empty import for new module 1`] = `
"import React from 'react';

import "babel/polyfill";

export default () => <div>Hello world</div>;"
`;

exports[`inserts named import for existing module 1`] = `
"
    import React, { Component } from 'react';

    export default () => <div>Hello world</div>;
  "
`;

exports[`inserts named import for new module 1`] = `
"import React from 'react';

import { Components } from "exponent";

export default () => <div>Hello world</div>;"
`;

exports[`test does not insert duplicate default import for existing module 1`] = `
"
    import React, { Component } from \'react\';

    export default () => <div>Hello world</div>;
  "
`;

exports[`test does not insert duplicate empty import for new module 1`] = `
"
    import { Component } from \'react\';
    import \'babel/polyfill\';

    export default () => <div>Hello world</div>;
  "
`;

exports[`test does not insert duplicate named import for existing module 1`] = `
"
    import { Component } from \'react\';

    export default () => <div>Hello world</div>;
  "
`;

exports[`test inserts default import for existing module 1`] = `
"
    import React, { Component } from \'react\';

    export default () => <div>Hello world</div>;
  "
`;

exports[`test inserts default import for new module 1`] = `
"import React from \'react\';

import lodash from \"lodash\";

export default () => <div>Hello world</div>;"
`;

exports[`test inserts empty import for new module 1`] = `
"import React from \'react\';

import \"babel/polyfill\";

export default () => <div>Hello world</div>;"
`;

exports[`test inserts named import for existing module 1`] = `
"
    import React, { Component } from \'react\';

    export default () => <div>Hello world</div>;
  "
`;

exports[`test inserts named import for new module 1`] = `
"import React from \'react\';

import { Components } from \"exponent\";

export default () => <div>Hello world</div>;"
`;
